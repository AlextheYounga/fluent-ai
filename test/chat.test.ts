import { test, expect } from "bun:test";
import {
  openai,
  google,
  ollama,
  systemPrompt,
  userPrompt,
  tool,
  anthropic,
  fireworks,
  requestObject,
} from "../src";
import { z } from "zod";

// prettier-ignore
const jobs = [
  openai({ apiKey: "<key>" }).chat("gpt-4o-mini"),
  google({ apiKey: "<key>" }).chat("gemini-1.5-flash"),
  anthropic({ apiKey: "<key>" }).chat("claude-3-5-sonnet-20241022"),
  ollama().chat("llama3.2"),
  fireworks({apiKey: "<key>"}).chat("accounts/fireworks/models/llama-v3p1-70b-instruct"),
];

test("chat", async () => {
  for (const job of jobs) {
    expect(
      await requestObject(
        job
          .messages([
            systemPrompt("you are a helpful assistant"),
            userPrompt("hi"),
          ])
          .temperature(0.5)
          .makeRequest()
      )
    ).toMatchSnapshot();
  }
});

test("json_object", async () => {
  expect(
    await requestObject(
      openai({ apiKey: "<key>" })
        .chat("gpt-4o-mini")
        .messages([userPrompt("hi")])
        .responseFormat({ type: "json_object" })
        .makeRequest()
    )
  ).toMatchSnapshot();
});

test("function calling", async () => {
  const weatherTool = tool("get_current_weather")
    .description("Get the current weather in a given location")
    .parameters({
      type: "object",
      properties: {
        location: {
          type: "string",
          description: "The city and state, e.g. San Francisco, CA",
        },
        unit: {
          type: "string",
          enum: ["celsius", "fahrenheit"],
        },
      },
      required: ["location"],
    });

  for (const job of jobs) {
    expect(
      await requestObject(
        job
          .tool(weatherTool)
          .messages([userPrompt("What's the weather like in Boston today?")])
          .makeRequest()
      )
    ).toMatchSnapshot();
  }
});

test("zod", async () => {
  const weatherTool = tool("get_current_weather")
    .description("Get the current weather in a given location")
    .parameters(
      z.object({
        location: z.string(),
        unit: z.enum(["celsius", "fahrenheit"]).optional(),
      })
    );

  for (const job of jobs) {
    expect(
      await requestObject(
        job
          .tool(weatherTool)
          .messages([userPrompt("What's the weather like in Boston today?")])
          .makeRequest()
      )
    ).toMatchSnapshot();
  }
});

test("jsonSchema", async () => {
  const personSchema = z.object({
    name: z.string(),
    age: z.number(),
  });

  for (const job of jobs) {
    expect(
      await requestObject(
        job
          .messages([
            userPrompt("generate a person with name and age in json format"),
          ])
          .jsonSchema(personSchema)
          .makeRequest()
      )
    ).toMatchSnapshot();
  }
});
