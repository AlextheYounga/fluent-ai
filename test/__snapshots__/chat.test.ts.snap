// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`chat 1`] = `
{
  "body": "{"messages":[{"role":"system","content":"you are a helpful assistant"},{"role":"user","content":"hi"}],"model":"gpt-4o-mini","temperature":0.5}",
  "headers": {
    "authorization": "Bearer <key>",
    "content-type": "application/json",
  },
  "method": "POST",
  "url": "https://api.openai.com/v1/chat/completions",
}
`;

exports[`chat 2`] = `
{
  "body": "{"contents":[{"role":"model","parts":[{"text":"you are a helpful assistant"}]},{"role":"user","parts":[{"text":"hi"}]}]}",
  "headers": {
    "content-type": "application/json",
  },
  "method": "POST",
  "url": "https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key=%3Ckey%3E",
}
`;

exports[`chat 3`] = `
{
  "body": "{"model":"claude-3-5-sonnet-20241022","messages":[{"role":"system","content":"you are a helpful assistant"},{"role":"user","content":"hi"}]}",
  "headers": {
    "anthropic-version": "2023-06-01",
    "content-type": "application/json",
    "x-api-key": "<key>",
  },
  "method": "POST",
  "url": "https://api.anthropic.com/v1/messages",
}
`;

exports[`chat 4`] = `
{
  "body": "{"model":"llama3.2","messages":[{"role":"system","content":"you are a helpful assistant"},{"role":"user","content":"hi"}],"tools":[],"stream":false}",
  "headers": {},
  "method": "POST",
  "url": "http://localhost:11434/api/chat",
}
`;

exports[`chat 5`] = `
{
  "body": "{"messages":[{"role":"system","content":"you are a helpful assistant"},{"role":"user","content":"hi"}],"model":"accounts/fireworks/models/llama-v3p1-70b-instruct","temperature":0.5}",
  "headers": {
    "authorization": "Bearer <key>",
    "content-type": "application/json",
  },
  "method": "POST",
  "url": "https://api.fireworks.ai/inference/v1/chat/completions",
}
`;

exports[`json_object 1`] = `
{
  "body": "{"messages":[{"role":"user","content":"hi"}],"model":"gpt-4o-mini","response_format":{"type":"json_object"}}",
  "headers": {
    "authorization": "Bearer <key>",
    "content-type": "application/json",
  },
  "method": "POST",
  "url": "https://api.openai.com/v1/chat/completions",
}
`;

exports[`function calling 1`] = `
{
  "body": "{"messages":[{"role":"user","content":"What's the weather like in Boston today?"}],"model":"gpt-4o-mini","temperature":0.5,"tools":[{"type":"function","function":{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string","description":"The city and state, e.g. San Francisco, CA"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"]}}}]}",
  "headers": {
    "authorization": "Bearer <key>",
    "content-type": "application/json",
  },
  "method": "POST",
  "url": "https://api.openai.com/v1/chat/completions",
}
`;

exports[`function calling 2`] = `
{
  "body": "{"contents":[{"role":"user","parts":[{"text":"What's the weather like in Boston today?"}]}]}",
  "headers": {
    "content-type": "application/json",
  },
  "method": "POST",
  "url": "https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key=%3Ckey%3E",
}
`;

exports[`function calling 3`] = `
{
  "body": "{"model":"claude-3-5-sonnet-20241022","messages":[{"role":"user","content":"What's the weather like in Boston today?"}],"tools":[{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string","description":"The city and state, e.g. San Francisco, CA"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"]}}]}",
  "headers": {
    "anthropic-version": "2023-06-01",
    "content-type": "application/json",
    "x-api-key": "<key>",
  },
  "method": "POST",
  "url": "https://api.anthropic.com/v1/messages",
}
`;

exports[`function calling 4`] = `
{
  "body": "{"model":"llama3.2","messages":[{"role":"user","content":"What's the weather like in Boston today?"}],"tools":[{"type":"function","function":{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string","description":"The city and state, e.g. San Francisco, CA"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"]}}}],"stream":false}",
  "headers": {},
  "method": "POST",
  "url": "http://localhost:11434/api/chat",
}
`;

exports[`function calling 5`] = `
{
  "body": "{"messages":[{"role":"user","content":"What's the weather like in Boston today?"}],"model":"accounts/fireworks/models/llama-v3p1-70b-instruct","temperature":0.5,"tools":[{"type":"function","function":{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string","description":"The city and state, e.g. San Francisco, CA"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"]}}}]}",
  "headers": {
    "authorization": "Bearer <key>",
    "content-type": "application/json",
  },
  "method": "POST",
  "url": "https://api.fireworks.ai/inference/v1/chat/completions",
}
`;

exports[`zod 1`] = `
{
  "body": "{"messages":[{"role":"user","content":"What's the weather like in Boston today?"}],"model":"gpt-4o-mini","temperature":0.5,"tools":[{"type":"function","function":{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string","description":"The city and state, e.g. San Francisco, CA"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"]}}},{"type":"function","function":{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"],"additionalProperties":false,"$schema":"http://json-schema.org/draft-07/schema#"}}}]}",
  "headers": {
    "authorization": "Bearer <key>",
    "content-type": "application/json",
  },
  "method": "POST",
  "url": "https://api.openai.com/v1/chat/completions",
}
`;

exports[`zod 2`] = `
{
  "body": "{"contents":[{"role":"user","parts":[{"text":"What's the weather like in Boston today?"}]}]}",
  "headers": {
    "content-type": "application/json",
  },
  "method": "POST",
  "url": "https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key=%3Ckey%3E",
}
`;

exports[`zod 3`] = `
{
  "body": "{"model":"claude-3-5-sonnet-20241022","messages":[{"role":"user","content":"What's the weather like in Boston today?"}],"tools":[{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string","description":"The city and state, e.g. San Francisco, CA"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"]}},{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"],"additionalProperties":false,"$schema":"http://json-schema.org/draft-07/schema#"}}]}",
  "headers": {
    "anthropic-version": "2023-06-01",
    "content-type": "application/json",
    "x-api-key": "<key>",
  },
  "method": "POST",
  "url": "https://api.anthropic.com/v1/messages",
}
`;

exports[`zod 4`] = `
{
  "body": "{"model":"llama3.2","messages":[{"role":"user","content":"What's the weather like in Boston today?"}],"tools":[{"type":"function","function":{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string","description":"The city and state, e.g. San Francisco, CA"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"]}}},{"type":"function","function":{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"],"additionalProperties":false,"$schema":"http://json-schema.org/draft-07/schema#"}}}],"stream":false}",
  "headers": {},
  "method": "POST",
  "url": "http://localhost:11434/api/chat",
}
`;

exports[`zod 5`] = `
{
  "body": "{"messages":[{"role":"user","content":"What's the weather like in Boston today?"}],"model":"accounts/fireworks/models/llama-v3p1-70b-instruct","temperature":0.5,"tools":[{"type":"function","function":{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string","description":"The city and state, e.g. San Francisco, CA"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"]}}},{"type":"function","function":{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"],"additionalProperties":false,"$schema":"http://json-schema.org/draft-07/schema#"}}}]}",
  "headers": {
    "authorization": "Bearer <key>",
    "content-type": "application/json",
  },
  "method": "POST",
  "url": "https://api.fireworks.ai/inference/v1/chat/completions",
}
`;

exports[`jsonSchema 1`] = `
{
  "body": "{"messages":[{"role":"user","content":"generate a person with name and age in json format"}],"model":"gpt-4o-mini","temperature":0.5,"response_format":{"type":"json_schema","json_schema":{"schema":{"type":"object","properties":{"name":{"type":"string"},"age":{"type":"number"}},"required":["name","age"],"additionalProperties":false,"$schema":"http://json-schema.org/draft-07/schema#"}}},"tools":[{"type":"function","function":{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string","description":"The city and state, e.g. San Francisco, CA"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"]}}},{"type":"function","function":{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"],"additionalProperties":false,"$schema":"http://json-schema.org/draft-07/schema#"}}}]}",
  "headers": {
    "authorization": "Bearer <key>",
    "content-type": "application/json",
  },
  "method": "POST",
  "url": "https://api.openai.com/v1/chat/completions",
}
`;

exports[`jsonSchema 2`] = `
{
  "body": "{"contents":[{"role":"user","parts":[{"text":"generate a person with name and age in json format"}]}]}",
  "headers": {
    "content-type": "application/json",
  },
  "method": "POST",
  "url": "https://generativelanguage.googleapis.com/v1beta/models/gemini-1.5-flash:generateContent?key=%3Ckey%3E",
}
`;

exports[`jsonSchema 3`] = `
{
  "body": "{"model":"claude-3-5-sonnet-20241022","messages":[{"role":"user","content":"generate a person with name and age in json format"}],"tools":[{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string","description":"The city and state, e.g. San Francisco, CA"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"]}},{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"],"additionalProperties":false,"$schema":"http://json-schema.org/draft-07/schema#"}}]}",
  "headers": {
    "anthropic-version": "2023-06-01",
    "content-type": "application/json",
    "x-api-key": "<key>",
  },
  "method": "POST",
  "url": "https://api.anthropic.com/v1/messages",
}
`;

exports[`jsonSchema 4`] = `
{
  "body": "{"model":"llama3.2","messages":[{"role":"user","content":"generate a person with name and age in json format"}],"tools":[{"type":"function","function":{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string","description":"The city and state, e.g. San Francisco, CA"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"]}}},{"type":"function","function":{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"],"additionalProperties":false,"$schema":"http://json-schema.org/draft-07/schema#"}}}],"stream":false}",
  "headers": {},
  "method": "POST",
  "url": "http://localhost:11434/api/chat",
}
`;

exports[`jsonSchema 5`] = `
{
  "body": "{"messages":[{"role":"user","content":"generate a person with name and age in json format"}],"model":"accounts/fireworks/models/llama-v3p1-70b-instruct","temperature":0.5,"response_format":{"type":"json_schema","json_schema":{"schema":{"type":"object","properties":{"name":{"type":"string"},"age":{"type":"number"}},"required":["name","age"],"additionalProperties":false,"$schema":"http://json-schema.org/draft-07/schema#"}}},"tools":[{"type":"function","function":{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string","description":"The city and state, e.g. San Francisco, CA"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"]}}},{"type":"function","function":{"name":"get_current_weather","description":"Get the current weather in a given location","parameters":{"type":"object","properties":{"location":{"type":"string"},"unit":{"type":"string","enum":["celsius","fahrenheit"]}},"required":["location"],"additionalProperties":false,"$schema":"http://json-schema.org/draft-07/schema#"}}}]}",
  "headers": {
    "authorization": "Bearer <key>",
    "content-type": "application/json",
  },
  "method": "POST",
  "url": "https://api.fireworks.ai/inference/v1/chat/completions",
}
`;
